### DISCUSSION TITLE
Migrate from travis to github actions

### DISCUSSION THREAD

**Discussion by akhilerm** [openebs/openebs#3299] [General] [answered]
Currently all the openebs repositories use a combination of travis and github actions to build/test/push the container images. Eventually we need to move all the repositories from using travis to github actions.

operations that are done in travis:
- integration tests (using minikube)
- downstream repo tagging

Issues faced with github actions currently:
- not able to run minikube easily in github actions
- downstream repo  release script need to be changed to github actions from travis.

**Comment by kmova** [comment]
Till we full setup integration tests with the github actions, we will continue to make use of travis. 

To avoid travis and github actions conflicting with each other in pushing the images, here is a checklist for enabling multi-arch builds for an image. 

- [ ] Make sure the travis images are built and pushed with arch suffix
- [ ] For the release tagged images, push the latest tag of amd64 build into quay.io 
- [ ] Setup credentials on github actions to push multi-arch images to docker hub

**Reply by shubham14bajpai** [reply]
If we add arch suffix to the images which didn't have the earlier, this will require an upgrade for old pools and volumes which have deployments with older images.

**Reply by akhilerm** [reply]
@shubham14bajpai Can you link to the issues and PRs that you have created as part of the migration from travis to github actions?

**Comment by shubham14bajpai** [comment]
Issues and PRs for github actions
cstor-operators: https://github.com/openebs/cstor-operators/pull/223
cstor-csi: https://github.com/openebs/cstor-csi/pull/140
jiva-operators: https://github.com/openebs/jiva-operator/pull/28
upgrade: https://github.com/openebs/upgrade/issues/90

**Reply by akhilerm** [reply]
node-disk-manager : https://github.com/openebs/node-disk-manager/pull/548

**Reply by akhilerm** [reply]
LVM Local PV: https://github.com/openebs/lvm-localpv/pull/27

**Reply by akhilerm** [reply]
Dynamic Local PV: https://github.com/openebs/dynamic-localpv-provisioner/pull/32. This PR also removes travis from the repository

**Reply by shubham14bajpai** [reply]
Zfs localpv https://github.com/openebs/zfs-localpv/pull/293

**Reply by shubham14bajpai** [reply]
velero-plugin: https://github.com/openebs/velero-plugin/pull/149

**Reply by akhilerm** [reply]
cstor: https://github.com/openebs/cstor/pull/330, The PR migrates tests and also handles downstream release tagging

**Reply by akhilerm** [reply]
libcstor: https://github.com/openebs/libcstor/pull/86

**Comment by shubham14bajpai** [comment]
**Findings:** Unlike travis which takes master as the current branch on PRs github actions use a specific branch called `merge` while executing the workflows
```
  Run make all.amd64 || exit 1;
  make all.amd64 || exit 1;
  shell: /bin/bash -e {0}
  env:
    MINIKUBE_HOME: /home/runner/work/_temp
    TAG: ci
    BRANCH: merge
--> cspc-operator image <--openebs/cspc-operator-amd64:ci
----------------------------
Building for merge-dev VERSION
==> Removing old directory...
==> Building cspc-operator using go version go1.15.8 linux/amd64... 
```

**Reply by shubham14bajpai** [reply]
We can use the github.base_ref to set tags the images as the version depends on the branch names in different branches.
ref: https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context

**Comment by akhilerm** [comment]
The integration tests in github actions will use a separate build from the same source. This is because the build and test runs in separate jobs and we need to find a reliable way to transfer multiarch images from one job to another. Once this method for transferring artifacts is fixed, the following is the flow that we expect to achieve

Job 1,2,3... (The jobs where multiplatform images are built)
1. Build images
2. Push to local regsitry

Job X (integration tests)
1. Run minikube
2. Fetch images from local registry
3. Run integration tests using images from local registry
4. If test passes , push the images to dockerhub , quay etc.

**Comment by shubham14bajpai** [comment]
With github action moving to ubuntu 20.04 the iscsi_tcp modules are not loaded by default. We need to load the modules manually via modprobe to run the volumes.

**Comment by akhilerm** [comment, marked as answer]
Integration tests for all components has been now migrated to github actions.
List of repositories and associated PRs in which migration changes are added:
- cstor-operators: openebs/cstor-operators#223
- cstor-csi: openebs/cstor-csi#140
- jiva-operators: openebs/jiva-operator#28
- upgrade: openebs/upgrade#90
- node-disk-manager : openebs/node-disk-manager#548
- LVM Local PV: openebs/lvm-localpv#27
- Dynamic Local PV: openebs/dynamic-localpv-provisioner#32
- Zfs localpv openebs/zfs-localpv#293
- cstor: openebs/cstor#330
- velero-plugin: openebs/velero-plugin#149
- libcstor: openebs/libcstor#86
